src/data/viteDataPlugin.ts

```diff
@@ -0,0 +1,50 @@
+/**
+ * This is the entry point for the Data Plugin that reads the YAML Files defining tricks and combos
+ */
+
+import viteGetAllCombos from "./viteComboFetcher";
+import generateJsonSchemas from "./viteSchemaGenerator";
+import viteGetAllTricks from "./viteTrickFetcher"
+import {createHash} from "crypto"
+
+async function generateVirtualModuleContent() {
+  const tricks = await viteGetAllTricks();
+  const combos = await viteGetAllCombos(tricks)
+
+  // *Should* get reproducible results under Node 
+  // As this code is not run on the Browser, this should be no issue.
+  const contentToHash = JSON.stringify({tricks, combos})
+
+
+  const hash = createHash("md5").update(contentToHash).digest("hex")
+
+
+  return `
+  export const tricks = ${JSON.stringify(tricks)};
+  export const combos = ${JSON.stringify(combos)};
+  export const hash = "${hash}";
+
+  export default ({
+    tricks, combos, hash
+  })
+  `
+}
+
+
+
+export function viteDataPlugin() {
+  generateJsonSchemas()
+  return {
+    name: "vite-plugin-highline-freestyle-data",
+    resolveId(id: string): string | void {
+      if(id === "virtual:highline-freestyle-data") {
+        return "\0"+id
+      }
+    },
+    async load(id: string): Promise<string | void> {
+      if(id === "\0virtual:highline-freestyle-data") {
+        return await generateVirtualModuleContent()
+      }
+    }
+  }
+}
```

Comment:
I don't quite understand the return type of this function. You are returning an object that has a field `name` and then two functions? Don't you have to provide the `function` keyword to declare functions?

Reply:
Nope :) JS also allows you to use an arrow function (="lambda" function).

Funnily enough they work *slightly* different. It doesnt matter in this case, but I just thought it would be worth mentioning...

`function(){}` creates a new `this` context, while `()=>{}` just uses the parent's context.

What I mean by this:
```js
#!/usr/bin/env node

const a = {
  x: "bla",
  fn: function() {
    console.log("a:",this)
  }
}

const b = {
  x: "ble",
  fn: () => {
    console.log("b:",this)
  }
}

a.fn()
b.fn()
```

Will log
```
a: { x: 'bla', fn: [Function: fn] }
b: undefined
```
For a, `this` is the object itself,
For b its undefined, because the parent scope does not have `this` specified.

Reply:
So `resolveId` and `load` are lambda functions with names?

What exactly happens with the two functions, are they just callable from whatever imports `viteDataPlugin`? Or are they executed immediately?

Reply:
They are function definitions stored in the object that gets returned.

That object defines a vite plugin; it is consumed by the vite runtime. Whenever vite has an import to deal with, it sends that import to all plugins' `resolveId`. That function can then either return void / undefined, or return a resolved ID, in which case the Plugin takes over handling the import.

same thing happens with `load`. Can return undefined / void, or a string.